### PR TITLE
Add in ability populate defaults from sub Schemas

### PR DESCRIPTION
**Pull request by paetling**
review @tleach  if the 'populate_default_subdoc' is set to True, this will search into the subschema and create a document with the defaults define in that schema

### CODE REVIEW

File: schemer/__init__.py [hunk 1]
```diff
@@ -54,6 +54,10 @@ def apply_defaults(self, instance):
                 else:
                     instance[field] = default
 
+            elif isinstance(spec['type'], Schema) and 'populate_default_subdoc' in spec and spec['populate_default_subdoc']:
+                instance[field] = {}
+                field_type = spec['type']
+                field_type.apply_defaults(instance[field])
```
Comment (paetling) [inline schemer/__init__.py]: reorder structure of this loop to allow for easier extensibility and to clean up method

File: tests/schemer/sample.py [hunk 2]
```diff
@@ -111,3 +111,26 @@ def valid_doc(overrides=None):
...
+                                 'name': {'type': basestring, 'default': 'james'},
+                                 'birthdate': {'type': Schema({
+                                                               'month': {'type': basestring, 'default': 'april'},
+                                                               'day': {'type': basestring, 'default': '25'},
+                                                               'year': {'type': basestring, 'default': '1991'}})},
+                                 'education': {'type': Schema({
+                                                               'highschool_name': {'type': basestring, 'default': 'Harvard Prep'},
+                                                               'college_name': {'type': basestring, 'default': 'Harvard'}
+                                                               }),
+                                               'populate_default_subdoc': False}})
```
Comment (tleach) [inline tests/schemer/sample.py]: As per the other PR, can you try and amalgamate the sample use case for this feature into the `BlogPost` example. Maybe make the `meta` field have a default?

File: tests/schemer/schema_test.py [hunk 3]
```diff
@@ -351,3 +351,25 @@ def test_default_value_does_not_overwrite_existing(self):
...
+    def test_populate_semi_fille_document(self):
+        doc = {'country': {'phone_extension': '1',
+                           'primary_flag_color': 'red'},
+               'name': 'james'}
+        expected_doc = {'country': {'phone_extension': '1',
+                                    'primary_flag_color': 'red',
+                                    'address': {'latitude': '10.0.0'}},
+                        'name': 'james'}
+        schema_defaults_object.apply_defaults(doc)
+        self.assertEquals(doc, expected_doc)
```
Comment (tleach) [inline tests/schemer/schema_test.py]: Group these with the other default application tests.

**Comment by tleach** [comment]
Don't forget to update the readme.

File: README.md [hunk 4]
```diff
@@ -228,15 +228,28 @@ Bear in mind that using dynamic type functions in this way effectively defers th
...
 ```python
 schema = Schema({"num_wheels": {"type": int, "default": 4}})
 ```
-or as a reference to parameterless function which will be called at the point the default is applied:
+* It can be specified as a reference to parameterless function which will be called at the point the default is applied:
 ```python
 import datetime
 schema = Schema({"created_date": {"type": datetime, "default": datetime.utcnow}})
 ```
+* You can use defaults with Arrys and Schemas
```
Comment (tleach) [inline README.md]: *Arrays

File: schemer/__init__.py [hunk 5]
```diff
@@ -42,19 +48,9 @@ def apply_defaults(self, instance):
                     for item in value:
                         field_type.contained_type.apply_defaults(item)
 
-
                 # Bailout as we don't want to apply a default
                 continue
```
Comment (tleach) [inline schemer/__init__.py]: I don't think you need this any more.

File: schemer/__init__.py [hunk 6]
```diff
@@ -118,7 +114,7 @@ def _verify_type(self, spec, path):
 
         if isinstance(field_type, Schema):
             # Nested documents cannot have defaults or validation
```
Comment (tleach) [inline schemer/__init__.py]: This comment needs updating

File: schemer/__init__.py [hunk 7]
```diff
@@ -129,6 +125,8 @@ def _verify_type(self, spec, path):
         elif not isinstance(field_type, type) and not isinstance(field_type, types.FunctionType):
             raise SchemaFormatException("Unsupported field type at {}. Type must be a type, a function, an Array or another Schema", path)
 
+    def _valid_schema_default(self, value):
+        return value == {}
```
Comment (tleach) [inline schemer/__init__.py]: I don't think we want to necessarily restrict people to empty dict. I'd say we just check that the value is of type `dict`.

Comment (paetling) [inline schemer/__init__.py]: fi we do not restrict it, we are opening ourselves up to a situation where someone can add values that are not valid through defaults

File: tests/schemer/schema_test.py [hunk 8]
```diff
@@ -92,6 +92,51 @@ def test_default_value_of_incorrect_type(self):
...
+            {'wheel':{'type':Schema({'size': {'type': int, 'default': 32},
+                                     'brand': {'type': basestring, 'default': 'firestone'}}),
+                      'default':'wrong'}},
+            'wheel')
+
+    def test_default_schema_value_of_incorrect_type_2(self):
+        self.assert_spec_invalid(
+            {'wheel':{'type':Schema({'size': {'type': int, 'default': 32},
+                                     'brand': {'type': basestring, 'default': 'firestone'}}),
+                      'default':{'wrong': True}}},
```
Comment (tleach) [inline tests/schemer/schema_test.py]: To reiterate: I'm not sure we should necessarily consider this invalid.

File: schemer/__init__.py (hunk 7 above)
Comment (tleach) [inline schemer/__init__.py]: Yes, but they can already do that now within an individual field definition - they can create a default and add validation which contradicts that default. At some point it's the dev's responsibility to not do stupid shit.  

I could see a couple of scenarios where defaults at this level might actually be useful - e.g. you're reusing a common schema for a sub doc but only want to apply a default field at one inclusion point.

Comment (paetling) [inline schemer/__init__.py]: ok.  can make that change

**Pull request merged**